Refactor transValidityInterval: era split, remove protocol branching,…

danbaruka opened the pull request:
… cleanup Alonzo/Conway logic - 2025-07-01 07:33:54

Refactor transValidityInterval to use era-specific logic: implement Conway-specific translation (no protocol branching, always strict upper bound), update Conway instances, and clean up Alonzo to always use pre-Conway logic. Remove protocol-version-based branching and obsolete helpers.

Commit:
Refactor transValidityInterval: era split, remove protocol branching, cleanup Alonzo/Conway logic - 2025-07-01 07:33:54

lehins reviewed (changes requested):
Your changes are all correct with respect to Alonzo and Babbage eras, but for Conway you are missing the new implementation of `transValidityInterval` that did the new style translation. In other words, you can't use the implementation of `transValidityInterval` from Alonzo in Conway, instead you need a new implementation of this function that has semantics of `hardforkConwayTranslateUpperBoundForPlutusScripts`
Other than that it looks good.

lehins commented:
Hey @danbaruka did you still want to finish this task. It's ok if you don't, I can pass this on to finish to one of my teammates, but if you do wanna finish it yourself that is ok too, just let me know. There is no rush.

danbaruka commented:
Hey thanks for your feedback, I'll definitely work on that and fix it as soon as possible.

I now understand.

lehins commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Era.hs:
```diff
@@ -96,5 +96,4 @@ type instance EraRule "UPEC" AlonzoEra = ShelleyUPEC AlonzoEra
 
 -- | Starting with protocol version 9, we translate the upper bound of validity
 -- interval correctly for Plutus scripts.
```
Haddock should also be removed, not just the function :)
```suggestion
```

danbaruka replied:
Well noted

Commit:
Implement Conway-specific transValidityInterval with open upper bound; remove outdated Alonzo comment - 2025-07-21 08:18:42

danbaruka commented:
Conway: Implement new transValidityInterval with open upper bound semantics; clean up Alonzo comment

- Implemented transValidityInterval in Conway era to always use an open (strict) upper bound for Plutus scripts, matching hardforkConwayTranslateUpperBoundForPlutusScripts semantics.
- Removed outdated Haddock comment regarding validity interval translation from Alonzo era.

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/TxInfo.hs:
```diff
@@ -762,3 +764,26 @@ class
+-- | Translate a validity interval to POSIX time for Conway era, always using an open upper bound (strictUpperBound).
+transValidityInterval ::
+  Inject (ConwayContextError era) a =>
+  Tx era ->
```
This function does not require it to be a `Tx`. It is safer to convert it to a `proxy`
```suggestion
  proxy era ->
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/TxInfo.hs:
```diff
@@ -762,3 +764,26 @@ class
+  ValidityInterval (SJust i) SNothing -> PV1.from <$> transSlotToPOSIXTime i
+  ValidityInterval SNothing (SJust i) -> do
+    t <- transSlotToPOSIXTime i
+    pure $ PV1.to t
```
Looks like you forgot to change the implementation to the expected Conway behavior:
```suggestion
    pure $ PV1.Interval
            (PV1.LowerBound PV1.NegInf True)
            (PV1.strictUpperBound t)
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/TxInfo.hs:
```diff
@@ -762,3 +764,26 @@ class
+
+-- | Translate a validity interval to POSIX time for Conway era, always using an open upper bound (strictUpperBound).
+transValidityInterval ::
+  Inject (ConwayContextError era) a =>
```
You'll need access to the `era` parameter
```suggestion
  forall proxy era a.
  Inject (ConwayContextError era) a =>
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/TxInfo.hs:
```diff
@@ -762,3 +764,26 @@ class
+  where
+    transSlotToPOSIXTime =
+      left (inject . BabbageContextError . TimeTranslationPastHorizon)
+        . slotToPOSIXTime epochInfo systemStart
```
In order to reduce duplication `transSlotToPOSIXTime` should be extracted into a standalone function in Alonzo and then reused here
```suggestion
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/TxInfo.hs:
```diff
@@ -20,6 +20,8 @@ module Cardano.Ledger.Conway.TxInfo (
   ConwayEraPlutusTxInfo (..),
   transTxBodyWithdrawals,
   transTxCert,
+  -- Era-specific translation for Plutus validity interval
```
This comment is unnecessary
```suggestion
```

lehins reviewed (changes requested):
There are still some issues. Also compilation warnings need to be addressed

Commit:
Refactor(txinfo): Fix Conway validity interval and reuse time translation logic - 2025-07-22 12:06:56

danbaruka commented:
All requested changes have been addressed:
- `transValidityInterval` now uses `proxy era` and has the correct `forall` quantifier.
- The implementation uses the correct open upper bound logic for Conway.
- The redundant comment was removed.
- `transSlotToPOSIXTime` is now shared and reused to avoid duplication.

Let me know if anything else needs to be updated!

lehins commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Plutus/TxInfo.hs:
```diff
@@ -214,39 +215,43 @@ transLookupTxOut (UTxO utxo) txIn =
+        pure $ PV1.Interval (PV1.lowerBound PV1.NegInf) (PV1.strictUpperBound t)
+    | otherwise -> PV1.to <$> slotToTime i
+  ValidityInterval (SJust i) (SJust j)
+    | pvMajor pv >= natVersion @9 -> do
```
What's up with this protocol version? The whole reason why this refactoring was suggested precisely to avoid branching on protocol version. Also this function is never gonna be applied to protocol version 9.

danbaruka replied:
Well noted thank

lehins reviewed (changes requested):
Please, at the very least make sure your PR builds without any warnings and passes all the tests before asking for review.

Commit:
Conway: always use open upper bound in transValidityInterval; remove protocol version checks - 2025-07-23 13:53:00

lehins commented:
Closing in favor of a correctly implemented #5230
